datapreprocess/ray/tokenize_shuffle.py

```diff
@@ -1,40 +1,240 @@
+    if file_type == RawFileType.GZIP_JSONL_COMPRESSED:
+        return lambda x: gzip_compressed_reader(x, content_key=content_key)
+    else:
+        raise Exception("Unsupported filetype")
+
+
+def tokenize_tiktoken(tokenizer, string):
+    eot_token = "<|endoftext|>"
+    pad_token = "<|pad|>"
+    return tokenizer.encode(string, allowed_special={eot_token, pad_token})
+
+
+def tokenize_eleutherai(tokenizer, string):
+    return tokenizer(string).input_ids
+
+
+def get_raw_filetype(key: str):
+    if key.endswith(".jsonl") or key.endswith(".json"):
+        return RawFileType.JSONL
+    elif key.endswith(".jsonl.zst") or key.endswith(".json.zst"):
+        return RawFileType.ZSTD_JSONL_COMPRESSED
+    elif key.endswith(".jsonl.gz") or key.endswith(".json.gz"):
+        return RawFileType.GZIP_JSONL_COMPRESSED
+    else:
+        logger.warning(f"Unknown filetype: {key}")
+        return RawFileType.UNKNOWN
+
+
+def preprocess(
+    key: str,
+    fh: BinaryIO,
+    seed: int,
+    content_key: str,
+    seqlen: int = 8192,
+    tokenizer=None,
+    do_sample: bool = False,
+):
+    rng = random.Random(hash(key) + seed)
+    if do_sample:
+        sample_freq = Sources.get_sampling_frequency(key)
+    buffer = []
+    tokenizer_fn, vocab_size = tokenizer
+    try:
+        file_type = get_raw_filetype(key)
+        if file_type == RawFileType.UNKNOWN:
+            return []
+        file_reader = get_reader(file_type, content_key)
+        pbar = tqdm(file_reader(fh))
+        pbar.set_description(key)
+        for string in pbar:
+            tokens = tokenizer_fn(string)
+            EOT = SpecialTokens.END_OF_TEXT.value % (vocab_size + len(SpecialTokens))
+            tokens.append(EOT)
+            buffer += tokens
+            while len(buffer) >= seqlen:
+                if do_sample:
+                    local_sample_freq = sample_freq
+                    # This code does the following
+                    # yield a int(sample_freq) copies of buffer[:seqlen]
+                    # then yield 1 more sample with Pr[sample_freq - int(sample_freq)]
+                    # in expectation we will yield sample_freq copies of buffer[:seqlen]
+                    while local_sample_freq > 1:
+                        yield buffer[:seqlen]
+                        local_sample_freq -= 1
+                    if rng.random() < local_sample_freq:
+                        yield buffer[:seqlen]
+                    buffer = buffer[seqlen:]
+                else:
+                    yield buffer[:seqlen]
+                    buffer = buffer[seqlen:]
+
+    except (IncompleteReadError, ReadTimeoutError, ResponseStreamingError) as e:
+        logger.error(f"There was an incomplete read error: {e} for key {key}")
+        return
+
+
+def process_keys(data, tokenizer, seqlen, seed, content_key):
+    s3_client = boto3.client("s3")
+    path = data["path"]
+    bucket, key = parse_s3_path(path)
```

Comment:
This assumes that `path` is an s3 path, right? Maybe this should be changed (at least later) to support local fs.